### ISSUE TITLE
v3.0.0 without mcrypt

### ISSUE THREAD

**Issue by TobyRyuk** [onelogin/php-saml#255]
I'm currently using php-saml in a project where we need to upgrade the PHP version to 7.1 and saw that the dev branch for v3.0.0 has removed mcrypt which is deprecated in PHP 7.1

Are there any plans on when to release v3.0.0 or is there something left to do? Would love to keep on using php-saml since it has been a breeze!

**Comment by recipe** [comment]
👍

**Comment by pitbulk** [comment]
There is no official release date, we will add to the 3.0.0 release not only the PHP7.1 support + mcrypt sessionless, but also code refactor + UX improvements and new functionality that right now is not possible to back-compatibility with the 2.X branch.

**Comment by ralphschindler** [comment]
I can appreciate that 3.0 will solve the mcrypt issue in full, but in the mean time do you think it is possible we get a 2.x release of the php-saml library where `ext-mcrypt` is moved to the `suggest` section (out of require) so that we can install this library with php 7.2?  We do not use any of the assertion features, so the code path that utilizes calls to mcrypt are never exercised in our use case.

Thanks!

**Comment by pitbulk** [comment]
The [remove_mcrypt](https://github.com/onelogin/php-saml/tree/remove_mcrypt) branch is exactly that.

**Comment by ralphschindler** [comment]
I guess the problem we have with this is that `onelogin/php-saml` is actually a sub-dependency of one of our authentication components that is a top level dependency of an application project.  If the project's `minimum-stability` is `stable`, that means our authentication dependency is going to fail since the top level composer has not whitelisted the `onelogin/php-saml` version `dev-mcrypt_removed@dev` version.  (We have at this point 10's of php apps throughout the company that use onelogin saml authentication, and likely would rather not have to whitelist this package in all of them).

Do you know perhaps when you expect 3.0 to drop?

Regarding the `mcrypt_removed` branch, it looks like more than a change of the composer.json's `require` and `suggest` sections. Is it as stable as releases cut from `master`?

**Comment by pitbulk** [comment]
mcrypt-removed is not updated with the latest code version. 3.0.0 version was updated 2 days ago...it lacks namespace support that I plan to add soon but I asked for some feedback for people that started that task but didnt complete..and waiting for a reply.

That said.. 3.0.0 version will not be compatible with 2.X so will require changes on your apps.

**Comment by KarimGeiger** [comment]
👍  I'd really appreciate it if the removal of mcrypt (or the release of version 3) could be prioritized because it's the only dependency which prevents us from upgrading to PHP 7.2.

**Comment by darrenmothersele** [comment]
PHP 7.2 support would be really nice!

**Comment by tjallingt** [comment]
What is the status on this? is it possible to merge the remove_mcrypt branch into a v2 release early while we wait for the v3 to stabilize?

**Comment by pitbulk** [comment]
@tjallingt You can use the current 3.0.0 branch that is exactly that (remove_mcrypt + recent changes of v2 branch). The official 3.0.0 release will include also namespace support. I'm still working on it.

**Comment by tjallingt** [comment]
Great, thank you so much! if you need help with some specific feature/refactor i'd be willing to take a look 😄

**Comment by jbq** [comment]
Had to run `composer require onelogin/php-saml:3.0.0.x-dev` to make it work

**Comment by ralphschindler** [comment]
Similarly, this is less than ideal, but since we have a private package that our apps use (sub-dependency), our sub dependency is utilizing this trick:

```
    "require": {
        "onelogin/php-saml": "3.0.0.x-dev as 3.0.0"
    },
```

3.0, or at least stable mcryptless onelogin cannot come soon enough.

**Comment by daniel-samson** [comment]
Do you have any idea when version 3.0.0 will be released?

I have tested our product with the dev version, however I do not feel confident that we can use the dev version in our next release. Is there anyway we could have this change in a "stable" version?

**Comment by pitbulk** [comment]
I made some progress adding namespace support, but I need to adapt the test suite now.

**Comment by pitbulk** [comment]
Now is available for testing:
- [3.0.0](https://github.com/onelogin/php-saml/tree/3.0.0) Compatible with PHP 7.X and uses Namespaces Candidate to be the official php-saml 3.0.0 version.
- [3.0.0-nameless](https://github.com/onelogin/php-saml/tree/3.0.0-namespaceless) A version compatible with PHP 7.X, but no Namespaces. Available on the repository, but will not be officially published.

**Comment by ralphschindler** [comment]
This works for us pretty cleanly, we've had minimal changes to work with the namespaced version of the 3.0.0 branch.  @pitbulk what is your roadmap for 3.0 becoming stable?

**Comment by pitbulk** [comment]
I think this version 3.0.0 can be released once validated.

**Comment by rdallimore** [comment]
It works for us but only with the full composer classmap:

`composer require onelogin/php-saml:3.0.0.x-dev`

results in `Fatal error: Class 'OneLogin\Saml2\Auth' not found ...`

`composer install -o` fixes it

**Comment by stof** [comment]
The master branch (i.e. 2.x) contains lots of changes since the time 3.x was branched out of it. It may be worth bring these changes back in 3.x too before the release.

**Comment by pitbulk** [comment]
@stof I already included them in 3.X branch. Have you found some commit not added?

**Comment by kynx** [comment]
@pitbulk just migrated to 3.0.0.x-dev. Pretty painless code changes and my integration tests are green. Good job 👍

**Comment by tombutterworth** [comment]
@pitbulk Is the release of the official php-saml 3.0.0 version likely to be soon?

**Comment by tobyryuk** [comment]
Sorry for the super late response, tested/reviewed the new changes to 3.X and it works smoothly for our projects @pitbulk

**Comment by tjallingt** [comment]
@pitbulk sorry for the delay, my colleague tested this through https://github.com/aacotroneo/laravel-saml2/tree/remove_mcrypt everything works just perfectly.

I'd say :shipit:

**Comment by ralphschindler** [comment]
@pitbulk Just checking in, any idea when 3.0 will be stable?

**Comment by NicolasCARPi** [comment]
Hello, just pinging this issue as the mcrypt dependency is a major blocker for php 7.2! Is there any news?

**Comment by danielbachhuber** [comment]
As a heads up, `mcrypt` can be installed for PHP 7.2 via `pecl`:

```
pecl install channel://pecl.php.net/mcrypt-1.0.1
```

For a Docker example https://github.com/pantheon-systems/wp-saml-auth/pull/127/files#diff-29944324a3cbf9f4bd0162dfe3975d88R12

**Comment by danielbachhuber** [comment]
> that's good to know, but mcrypt has been removed for good reasons, so I'd love to be able to never use it anymore anywhere :p

Just to clarify, `mcrypt` has been removed because it's been abandoned for some time: https://wiki.php.net/rfc/mcrypt-viking-funeral

While yes, it'd be great to remove `mcrypt` from the equation, doing so today vs. three months from now is nominal given how long `mcrypt` is abandoned.

**Comment by pitbulk** [comment]
My initial intention was to official release 3.0.0 version with [prs-7](https://github.com/onelogin/php-saml/issues/162) but that implementation had no much progress, so I will officially release 3.0.0 without that feature.

**Comment by NicolasCARPi** [comment]
@pitbulk yes, thanks! Sometimes it's better to not try and cram too many features in a release ;)

**Comment by RinkAttendant6** [comment]
@pitbulk is there any timeline on releasing 3.0.0 without the PSR-7 feature? I just found this library and it looks quite promising, but the 2.x version doesn't work on PHP 7.2.

**Comment by pitbulk** [comment]
It will be released next week.

**Comment by mikekamornikov** [comment]
@pitbulk Is it deferred? 2 weeks passed

**Comment by pitbulk** [comment]
I'm porting this change to PHP
https://github.com/onelogin/java-saml/commit/ba141697df4bb1cd670c61307e1de7767670bf8c

and after that, I will release the new version. Sorry for the delay guys!.

**Comment by bytestream** [comment]
Bit late to the party but is there any documentation for upgrading from v2 to v3 @pitbulk ? To quote your earlier comment:
>  3.0.0 version will not be compatible with 2.X so will require changes on your apps.

Is it just namespaces?
